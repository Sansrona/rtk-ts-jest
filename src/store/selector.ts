import { RootState } from "./index";
export const selectTodos = (state: any) => state.todos;
