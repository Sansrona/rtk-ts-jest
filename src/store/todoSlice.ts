import {
  createSlice,
  createAsyncThunk,
  AnyAction,
  PayloadAction,
} from "@reduxjs/toolkit";

type Todo = {
  id: string;
  title: string;
  completed: boolean;
};

type TodosState = {
  list: Todo[];
  error: string | null;
  loading: boolean;
};

const initialState: TodosState = {
  list: [],
  error: null,
  loading: false,
};

export const fetchTodos = createAsyncThunk<
  Todo[],
  undefined,
  { rejectValue: string }
>("todos/fetchTodos", async function (_, { rejectWithValue }) {
  const response = await fetch(
    "https://jsonplaceholder.typicode.com/todos?_limit=10"
  );
  if (!response.ok) {
    return rejectWithValue("server error!");
  }

  return await response.json();
});

export const addNewTodo = createAsyncThunk<
  Todo,
  string,
  { rejectValue: string }
>("todos/addNewTodo", async function (text, { rejectWithValue }) {
  const todo = { userId: 1, title: text, completed: false };
  const response = await fetch("https://jsonplaceholder.typicode.com/todos", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(todo),
  });
  if (!response.ok) {
    return rejectWithValue("server error! can`t add a new task");
  }
  return await response.json();
});

export const toggleStatus = createAsyncThunk<
  Todo,
  string,
  { rejectValue: string; state: { todos: TodosState } }
>("todos/toggleStatus", async function (id, { rejectWithValue, getState }) {
  const todo = getState().todos.list.find((t) => t.id === id);

  if (todo) {
    const response = await fetch(
      `https://jsonplaceholder.typicode.com/todos/${id}`,
      {
        method: "PATCH",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          completed: !todo.completed,
        }),
      }
    );

    if (!response.ok) {
      return rejectWithValue("Can't toggle status. Server error.");
    }

    return (await response.json()) as Todo;
  }
  return rejectWithValue("No such todo in the list!");
});

export const deleteTodo = createAsyncThunk<
  string,
  string,
  { rejectValue: string }
>("todos/deleteTodo", async function (id, { rejectWithValue }) {
  const response = await fetch(
    `https://jsonplaceholder.typicode.com/todos/${id}`,
    {
      method: "DELETE",
    }
  );

  if (!response.ok) {
    return rejectWithValue("Can't delete task. Server error.");
  }

  return id;
});

const todo = createSlice({
  name: "todos",
  initialState,
  reducers: {},
  extraReducers(builder) {
    builder
      .addCase(fetchTodos.pending, (state) => {
        state.error = null;
        state.loading = true;
      })
      .addCase(fetchTodos.fulfilled, (state, action) => {
        state.loading = false;
        state.list = action.payload;
      })
      .addCase(addNewTodo.pending, (state) => {
        state.error = null;
      })
      .addCase(addNewTodo.fulfilled, (state, action) => {
        state.list.push(action.payload);
      })
      .addCase(toggleStatus.fulfilled, (state, action) => {
        const toggledTodo = state.list.find(
          (todo) => todo.id === action.payload.id
        );
        if (toggledTodo) {
          toggledTodo.completed = !toggledTodo.completed;
        }
      })
      .addCase(deleteTodo.fulfilled, (state, action) => {
        state.list = state.list.filter((todo) => todo.id !== action.payload);
      })
      .addMatcher(isError, (state, action: PayloadAction<string>) => {
        state.error = action.payload;
        state.loading = false;
      });
  },
});

export default todo.reducer;

const isError = (action: AnyAction) => {
  return action.type.endsWith("rejected");
};
