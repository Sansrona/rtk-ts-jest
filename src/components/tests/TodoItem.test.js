import * as reduxHooks from '../../hooks';
import * as actions from '../../store/todoSlice';
import { render, screen, fireEvent } from '@testing-library/react';

import TodoItem from '../TodoItem';

jest.mock('../../hooks');

const mockedUseAppDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');

describe('TodoItem', () => {
    it('should render a TodoItem', () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        const { view } = render(<TodoItem id='123' title='Redux' completed={false} />);

        expect(view).toMatchSnapshot();
    });
    it('should dispatch actions', () => {
        const dispatch = jest.fn();
        mockedUseAppDispatch.mockReturnValue(dispatch);
        const mockedToggleStatus = jest.spyOn(actions, 'toggleStatus');
        const mockedRemoveTodo = jest.spyOn(actions, 'deleteTodo');
        const { view } = render(<TodoItem id='123' title='Redux' completed={false} />);

        fireEvent.click(screen.getByRole('checkbox'));

        expect(dispatch).toHaveBeenCalledTimes(1);
        expect(mockedToggleStatus).toHaveBeenCalledWith('123');
        fireEvent.click(screen.getByRole('button'));
        expect(dispatch).toHaveBeenCalledTimes(2);
        expect(mockedRemoveTodo).toHaveBeenCalledWith('123');

    })
})