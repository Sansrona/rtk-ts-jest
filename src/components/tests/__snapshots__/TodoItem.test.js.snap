// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should dispatch actions 1`] = `undefined`;

exports[`TodoItem should render a TodoItem 1`] = `undefined`;
